### PR TITLE
first pass at putting types in IR comparisons

### PR DESCRIPTION
**Pull request by catoverdrive**
cc @tpoterba @cseed @danking

### CODE REVIEW

File: src/main/scala/is/hail/expr/ir/ComparisonOp.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package is.hail.expr.ir
+
+import is.hail.annotations.CodeOrdering
+import is.hail.asm4s.Code
+import is.hail.expr.types._
+import is.hail.utils.lift
+
+object ComparisonOp {
+
+  private def checkIncompatible[T](lt: Type, rt: Type): Unit =
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/ir/ComparisonOp.scala]: can we call this `checkCompatible`?

File: src/main/scala/is/hail/expr/ir/Interpret.scala [hunk 2]
```diff
@@ -183,6 +154,17 @@ object Interpret {
               }
             }
         }
+      case ApplyComparisonOp(op, l, r) =>
+        val lValue = interpret(l, env, args, agg)
+        val rValue = interpret(r, env, args, agg)
+        op match {
+          case EQ(_) => lValue == rValue
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/ir/Interpret.scala]: do these treat nan / infinity correctly?

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/Interpret.scala]:
```
scala> Float.NaN == Float.NaN
res0: Boolean = false

scala> Float.PositiveInfinity == Float.PositiveInfinity
res1: Boolean = true
```

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/Interpret.scala]: yeah. CodeOrdering should be implementing this definition of equality.

File: src/main/scala/is/hail/expr/AST.scala [hunk 3]
```diff
@@ -742,15 +742,22 @@ case class Apply(posn: Position, fn: String, args: Array[AST]) extends AST(posn,
...
+
   private[this] def tryIRConversion(agg: Option[(String, String)]): ToIRErr[IR] =
     for {
       irArgs <- all(args.map(_.toIR(agg)))
       ir <- fromOption(
         this,
         s"no function or primop found for $fn",
         tryPrimOpConversion(args.map(_.`type`).zip(irArgs)).orElse(
+          tryComparisonConversion(args.map(_.`type`).zip(irArgs)).orElse(
           IRFunctionRegistry.lookupConversion(fn, irArgs.map(_.typ))
```
Comment (danking) [inline src/main/scala/is/hail/expr/AST.scala]: We expect exactly one of these to match right? This is a good opportunity to create a utility function that asserts that behavior.

```scala
def exactlyOne(
  blame: IR,
  message: String,
  a: Option[T],
  b: Option[T],
  c: Option[T]
): ToIRErr[T] = (a, b, c) match {
  case (Some(x), None, None) => success(x)
  case (None, Some(x), None) => success(x)
  case (None, None, Some(x)) => success(x)
  case (x, y, z) => fail(blame, "expected exactly one of $x, $y, $z to be Some.", ToIRErr.getCaller())
}
```

I think the use site looks a little nicer too less `orElse` noise

```scala
    for {
      irArgs <- all(args.map(_.toIR(agg)))
      typedIRArgs = args.map(_.`type`).zip(irArgs)
      ir <- fromOption(
        this,
        s"no function or primop found for $fn",
        exactlyOneOf(
          tryPrimOpConversion(typedIRArgs),
          tryComparisonConversion(typedIRArgs),
          IRFunctionRegistry.lookupConversion(fn, irArgs.map(_.typ))
            .map { irf => irf(irArgs) })))
// ...
```

Also can we get away without `typedIRArgs`? the IR has the type now right?

File: src/main/scala/is/hail/expr/ir/BinaryOp.scala [hunk 4]
```diff
@@ -34,7 +25,8 @@ object BinaryOp {
   private def incompatible[T](lt: Type, rt: Type, op: BinaryOp): T =
     throw new RuntimeException(s"Cannot apply $op to $lt and $rt")
 
-  def emit(op: BinaryOp, lt: Type, rt: Type, l: Code[_], r: Code[_]): Code[_] = (lt, rt) match {
+  def emit(op: BinaryOp, lt: Type, rt: Type, l: Code[_], r: Code[_]): Code[_] =
+    (lt, rt) match {
```
Comment (danking) [inline src/main/scala/is/hail/expr/ir/BinaryOp.scala]: indentation is now wrong here

File: src/main/scala/is/hail/expr/ir/Emit.scala [hunk 5]
```diff
@@ -888,6 +906,13 @@ private class Emit(
   private def present(x: Code[_]): EmitTriplet =
     EmitTriplet(Code._empty, const(false), x)
 
+  private def strict(value: Code[_], args: EmitTriplet*): EmitTriplet = {
```
Comment (danking) [inline src/main/scala/is/hail/expr/ir/Emit.scala]: nice! let's use this with ApplyBinaryPrimOp too

File: src/test/scala/is/hail/expr/AstToIrSuite.scala [hunk 6]
```diff
@@ -77,26 +77,25 @@ class ASTToIRSuite extends TestNGSuite {
...
+    "1.0 != 2.0" -> ApplyComparisonOp(NEQ(TFloat64()), F64(1.0), F64(2.0)),
+    "0 // 0 + 1" -> ApplyBinaryPrimOp(
+      Add(),
+      ApplyBinaryPrimOp(RoundToNegInfDivide(), I32(0), I32(0)),
+      I32(1)),
+    "0 // 0 * 1" -> ApplyBinaryPrimOp(
+      Multiply(),
+      ApplyBinaryPrimOp(RoundToNegInfDivide(), I32(0), I32(0)),
+      I32(1))
     )
```
Comment (danking) [inline src/test/scala/is/hail/expr/AstToIrSuite.scala]: Is this parenthesis not aligned correctly now?

File: src/main/scala/is/hail/expr/ir/ComparisonOp.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package is.hail.expr.ir
+
+import is.hail.annotations.CodeOrdering
+import is.hail.asm4s.Code
+import is.hail.expr.types._
+import is.hail.utils.lift
+
+object ComparisonOp {
```
Comment (danking) [inline src/main/scala/is/hail/expr/ir/ComparisonOp.scala]: Can you opine a bit on why we're divorcing comparisons from binary primitive operations? We could also implement this by teaching binary primitive operations the distinction between strict- and lazy-missingness.

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/ComparisonOp.scala]: because i did a dumb. I'll move them back.

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/ComparisonOp.scala]: actually, the interface for emitting code was kind of hard to unify here, so i might leave as is.

**Pull request merged**